Fix #3469: Introduce PredictionAlgorithmRegistryService and a sample prediction service.

pranavsid98 opened the pull request:
#3469 

This PR introduces a very preliminary version of the Registry service.

Tests need to be added and then AnswerClassificationService needs to be modified to use this new framework.

@prasanna08 @anmolshkl 
Is this more or less the rough idea to implement this?

pranavsid98 committed:
first draft

anmolshkl commented:
@pranavsid98 I fear I won't be able to help with this. @AllanYangZhou Can you please check this?

seanlip commented:
@pranavsid98 -- I think this is actually a bit too little to go on. What would be helpful is to get a sense of how the learner view is expected to hook into this -- without that it's hard to tell whether this is good or not, in isolation. (We've had similar issues with previous PRs, as you know :) )

Would you like to add a bit more? You can have this return a mock result for the time being, but maybe try to get it to the point where it at least shows the whole flow when the learner submits an answer.

prasanna08 commented on core/templates/dev/head/pages/exploration_player/PredictionAlgorithmRegistryService.js:
```diff
@@ -0,0 +1,31 @@
+ * @fileoverview Service for mapping algorithmId to PredictionAlgorithmService.
+ */
+
+ oppia.factory('PredictionAlgorithmRegistryService', [
```
@pranavsid98 factory basically returns an object from this function and you can only use those functions which are defined in that returned object. So you should return something from here. (Also in JS objects are written as python dict. They keys are attributes of object and value can be anything, other object, function or a literal)

pranavsid98 replied:
Done.

pranavsid98 committed:
changes

pranavsid98 commented:
@seanlip I understand. I already have the whole flow in mind, but I wanted to test the individual components before I implemented the whole flow (I'm still sort of getting around on how code is structured in Angular). I have added some more components to the flow, but I'll explain the whole flow I have in mind here (@anmolshkl @prasanna08 @AllanYangZhou ):

1.  In core.controllers.reader, ExplorationHandler provides the data of the exploration to the front end. In this controller, I plan to attach a field to the State object (classifier_data) and pass the updated exploration domain object to the front end.

2. When a learner submits an answer, we trigger AnswerClassificationService.getMatchingClassificationResultNew (I haven't removed the older function since I wanted the new framework to work perfectly before doing that). This function in turn retrieves the algorithmId using a service in app.js - algorithmIdRetrievalService.  It then uses PredictionAlgorithmRegistryService to retrieve the prediction service. This service's predict() method is used to get the prediction result which is returned.

TODO now:

- Add tests for all services implemented (I'm still working on front end tests).

- Modify ExplorationHandler controller to implement point 1 above.

- Modify the structure of the data returned by predict() method to better fit its purpose.

PS: I've got some listing issues which I'm not quite sure about as well, but that can be resolved later I presume.
Thanks!

pranavsid98 commented:
@seanlip @prasanna08 Oh, Im slightly unsure about how it changes when I remove the async method as well. I've got a couple of return statements where I'm not entirely sure what to do - I've mentioned you in the comments there. Me and Prasanna decided that the new framework does not require async calls, so I was trying to remove them.

AllanYangZhou commented on core/templates/dev/head/app.js:
```diff
@@ -1220,3 +1220,20 @@ oppia.factory('codeNormalizationService', [function() {
+// Service for retrieving algorithm id of classifier for a given interaction id.
+oppia.factory('algorithmIdRetrievalService', [function() {
+  // Maintain the Interaction-Classifier mapping as defined in feconf.
+  var interactionClassifierMapping = {
```
Should this be kept as a constant somewhere else, or is there a reason to keep this here?

pranavsid98 replied:
@AllanYangZhou Is there a particular place to keep constants like these in the front end? I am a bit unsure about the structure of Oppia's front end (first time working with it).

prasanna08 replied:
@pranavsid98 would it be better to pass algorithmId with classifier data itself? That way we can use appropriate algorithm using which classifier data is obtained. It also gives us backward compatibility and we can use old approach if we have old classifier data. @seanlip what do you think?

pranavsid98 replied:
@prasanna08 We could. So, do you suggest attaching a dict to the State object with the following structure: 

state.classifier_details = {
    'algorithm_id': <>,
    'classifier_data': <>
}

seanlip replied:
@prasanna08 -- Yes, I think that's fine. The algorithm ID is already stored with the job results model that contains the classifier data, right? So no additional retrievals are needed?

@pranavsid98 -- I think that makes sense, but it'd be good to get @prasanna08's feedback.

pranavsid98 replied:
Done.

AllanYangZhou commented on core/templates/dev/head/app.js:
```diff
@@ -1220,3 +1220,20 @@ oppia.factory('codeNormalizationService', [function() {
 }]);
+
+// Service for retrieving algorithm id of classifier for a given interaction id.
+oppia.factory('algorithmIdRetrievalService', [function() {
```
Is there a case where we just want the algorithmId, or can we just return the PredictionService?

Also, I'm a bit confused about the use of the words algorithm in some places and predictionservice in others. Is there a difference? If they're being used as synonyms, we should just stick to one.

pranavsid98 replied:
@AllanYangZhou In this case, we just want the algorithm_id. This is required because the algorithm_id is what we use to retrieve its corresponding service. Basically speaking, there is a one-to-one correspondence between algorithm_id and prediction services.

AllanYangZhou commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -151,6 +153,63 @@ oppia.factory('AnswerClassificationService', [
+          var predictionService = (
+            PredictionAlgorithmRegistryService.getPredictionService(
+              algorithmId));
+          var result = predictionService.predict(classifierData, answer);
```
Could just return this directly. Also if you are going to define a variable better to choose a different name since there's already another variable called `result`.

pranavsid98 replied:
Done.

AllanYangZhou reviewed (changes requested):
The structure looks OK, a few initial comments

AllanYangZhou commented:
@pranavsid98  Just to make sure I understand the rule of `getMatchingClassificationResultNew` so that I'm reviewing this properly:

* First we check if there's a custom outcome to the student answer under the interaction rules
* If it's only the default outcome, then we use the classifier to classify the answer into an existing answer group

Does that sound about right?

pranavsid98 commented:
@AllanYangZhou Sounds right. I didn't mess much with the existing framework. According to the existing framework, as you said, we check for a custom outcome under interaction rules. If its the default outcome, we used to ping the back-end for prediction. I just replaced that part where we ping the back-end with the new PredictionService method.

seanlip commented on core/templates/dev/head/app.js:
```diff
@@ -1220,3 +1220,20 @@ oppia.factory('codeNormalizationService', [function() {
 }]);
+
+// Service for retrieving algorithm id of classifier for a given interaction id.
+oppia.factory('algorithmIdRetrievalService', [function() {
```
Not sure why you need this factory. Store interactionClassifierMapping in assets/constants.js and remove it from feconf if it's meant to be the same constant in both the frontend and the backend.

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -151,6 +153,63 @@ oppia.factory('AnswerClassificationService', [
+          alertsService.addWarning(
+            'Something went wrong with the exploration: no ' +
+            'interactionRulesService was available.');
+          // @prasanna08 @seanlip. Not sure about this part.
```
@pranavsid98 This is the case when there is no explicit rules service associated with the interaction, which seems like an error. What is your specific question (and best guess at a solution), though? Without knowing that, it's hard to know how to respond.

pranavsid98 replied:
@seanlip If we are going with async method right now, we can return deferred.reject(). But, when we do eventually make it a synchronous process, this should probably raise the exception and just end the execution there. I assume we can't do an empty return. What would be a good equivalent for deferred.reject()?

seanlip replied:
throw Error(...)?

seanlip replied:
I mean, in normal operation, this shouldn't happen...

pranavsid98 replied:
@seanlip We don't need to throw error do we? Doesn't the alertService add the warning for us already?

seanlip replied:
That's the user-visible part. But you'll need to throw an error to avoid further code execution, I think.

pranavsid98 replied:
Makes sense. But since we aren't really focusing on changing this into a synchronous method right now, we can let it be for now. We can make a separate PR as you suggested to achieve the above.

seanlip replied:
Yep, sounds good!

seanlip commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -151,6 +153,63 @@ oppia.factory('AnswerClassificationService', [
+          return;
+        }
+
+        if (result.outcome === defaultOutcome &&
```
Maybe a better logic flow is:

do rule-based classification
if (result) return result

mlPredictionResult = predictionService.predict(...)
return mlPredictionResult || defaultResult;   // this returns classification result if it's not falsy, otherwise it returns defaultResult

pranavsid98 replied:
@seanlip After rule-based classification, 
if(result) will be True even for ML prediction. As long as the three conditions (defaultOutcome, interaction_id is trainable, ENABLE_ML_CLASSIFIERS) for prediction hold, we run the prediction service. If any of the above conditions are false, we just return the result obtained from rule-based classification. I'm not quite sure how you want to modify this flow.

seanlip replied:
Apologies, I misread the code originally. You are correct.

That said, could we pull out a couple of variables so it's easy to read?

```
        var ruleBasedOutcomeIsDefault = (result.outcome === defaultOutcome);
        var interactionIsTrainable = INTERACTION_SPECS[oldState.interaction.id].is_string_classifier_trainable;
        if (ruleBasedOutcomeIsDefault && interactionIsTrainable && ENABLE_ML_CLASSIFIERS) {
```

Also, btw: why "is_string_classifier_trainable"? Not all classifiers are "string classifiers".

pranavsid98 replied:
Made the variable changes. I suppose the is_string_classifier_trainable was the implementation done in the early stages. I'll make it - is_classifier_trainable?

seanlip replied:
How about is_interaction_trainable? Classifiers are obviously trainable.

pranavsid98 replied:
Makes sense. Yep. I'll change it to is_interaction_trainable then.

seanlip reviewed (commented):
Good start! Don't worry about changing async --> sync for now, if that would make things easier -- we can do that in a separate PR.

I took a quick glance through and left some comments. I suggest that you work next on finalizing the whole thing (so use actual services rather than sample service etc. and see if you can get it working end-to-end on devserver) -- the general outline seems reasonable to me.

pranavsid98 committed:
few changes

pranavsid98 committed:
add basic front end tests

pranavsid98 committed:
front end tests

pranavsid98 committed:
added back-end test for passing state_classifier_mapping

pranavsid98 committed:
advance the pipeline

pranavsid98 commented:
@seanlip @prasanna08 @AllanYangZhou 
With this commit, I've slightly gone forward with the end-to-end implementation. The current process is outlined below. PTAL.

- core.controllers.reader.ExplorationHandler now passes an additional parameter (state_classifier_mapping) to the front-end. The proper functioning of this additional parameter has been tested.

- PlayerServices.init() receives the exploration from the backend and creates domain objects. So, before creation, we attach an additional field on exploration.states (classifier_details). 

- StateObjectFactory has been modified accordingly to fit in this new field (classifierDetails) exclusively for the front end and StatesObjectFactorySpec has been modified accordingly.

Apart from this, there are multiple occurrences where we might need to modify the code to fit in this new structure, which is not done yet. I shall take a comprehensive sweep throughout the codebase and fix all occurrences in a later commit.

Thanks!

pranavsid98 committed:
fix editor_Test

pranavsid98 committed:
fix

pranavsid98 committed:
fix frontend tests partially

seanlip commented on core/controllers/reader.py:
```diff
@@ -213,6 +213,25 @@ def get(self, exploration_id):
+            if classifier_training_jobs[index] == None:
+                state_classifier_mapping[state_name] = None
+            else:
+                # Not making this a multi-get since we are deprecating
```
No, this won't do. We can't do individual GETs in a for loop; it takes too long. This will block merge.

Why don't you deprecate ClassifierDataModel first and then do this PR?

pranavsid98 replied:
@seanlip We thought we'll do this first since it blocks Prasanna's PR on frontend Code classifier service and his GSOC project in general. ClassifierDataModel deprecation wasn't technically a blocker and more of an optimisation.  But, if you're certain that we can't do GETs in a for loop, I think it might be better to get the deprecation done with.

seanlip replied:
Yes, we should avoid GETs in a for loop, so this does need to be rewritten.

However, I was just wondering -- in this PR, can't you do get_multi() on the job results models without deprecating the ClassifierDataModels? I.e., you could just ignore/bypass the latter for now, and deprecate them later? (But if I'm misunderstanding and you do need to get rid of them for this to work, then yep, we'll need to remove them.)

pranavsid98 replied:
@seanlip We are already doing get_multi() to get all jobs. After retrieving jobs, we retrieve ClassifierData models too. We need ClassifierData because we don't know the classifier_data yet. That is what we will be passing to the front-end. We plan to keep classifier_data as a field of the job. So, when we do that, the get_multi() on the job will be sufficient.

seanlip replied:
Ah, I see -- thanks for explaining!

OK, in that case, by the time this PR is merged, classifier_data does need to sit in the job model so that we only do a single get_multi() to get all classifier jobs, and nothing else. It is important that the learner view loads fast.

pranavsid98 replied:
Yep. Lets deprecate ClassifierData first then.

seanlip commented on core/templates/dev/head/pages/exploration_player/PlayerServices.js:
```diff
@@ -213,6 +213,11 @@ oppia.factory('oppiaPlayerService', [
           }
           loadedExploration.then(function(data) {
+            // Update exploration's states field.
+            for (var stateName in data.exploration.states) {
```
No, don't do this. You'll need to modify the functions explicitly, e.g. ExplorationObjectFactory.createFromBackendDict(data.exploration, data.state_classifier_mapping). Be careful to check if there are other instances of ExplorationObjectFactory.createFromBackendDict() that don't have access to this data -- if so, you might need to construct the state stuff in the backend instead and then pass it over.

pranavsid98 replied:
@seanlip There are other instances that do not have access to this data. Precisely why I thought I'll do it here since Player services is the only service that receives the exploration for the learner. The learner view is the only place where we ever need the state_classifier_mapping.

seanlip replied:
Ah, sorry, no. That still won't do. It's important to not modify the keys of objects/sub-objects, otherwise their definition becomes quite unclear.

Here's an alternative. If this is only used in the learner view, consider having a separate StateToClassifierMappingService that is only available in the learner view. The Exploration and State domain objects remain as they are. Just fetch the classifier data from this service as needed.

pranavsid98 replied:
@seanlip That makes a lot of sense. I think we can do that. I'll see how we can work on it.

seanlip reviewed (commented):
Hi @pranavsid98 -- I'm not going to review this PR in detail until you've got it fully working end-to-end, but I did take a brief cursory glance and have a couple of comments. Thanks!

pranavsid98 commented:
@seanlip The full flow is implemented (Not sure about how it works end-to-end because there a few errors which need to be debugged, but the system looks like this end-to-end):

1. When learner submits an answer, PlayerServices.submitAnswer() is triggered. This, in turn, triggers AnswerClassificationService.getMatchingClassificationResult().
2. We retrieve the algorithmId and classifierData using the StateClassifierMappingService which is initialised inside the init() of PlayerServices. The data for StateClassifierMappingService is sent from core.reader.ExplorationHandler.
3. If rule-based classification doesn't work and conditions for training are satisfied, getMatchingClassificationResult() triggers the PredictionAlgorithmRegistryService and retrieves the prediction algorithm service based on algorithmId.
4. We run predict() using classifierData, userText and return the result to the PlayerServices.

A few tests are failing, but otherwise, I think this is the structure I had in mind. 
@prasanna08 @AllanYangZhou PTAL.

pranavsid98 committed:
full structure

pranavsid98 committed:
state classifier mapping service

prasanna08 commented on core/controllers/reader_test.py:
```diff
@@ -188,6 +194,68 @@ def test_give_feedback_handler(self):
+            '%s/%s' % (feconf.EXPLORATION_INIT_URL_PREFIX, self.exp_id))
+        retrieved_state_classifier_mapping = exploration_dict[
+            'state_classifier_mapping']
+        print retrieved_state_classifier_mapping
```
No need for this?

pranavsid98 replied:
Done.

prasanna08 commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -125,27 +127,19 @@ oppia.factory('AnswerClassificationService', [
-              .is_string_classifier_trainable &&
+        var ruleBasedOutcomeIsDefault = (result.outcome == defaultOutcome);
+        var interactionIsTrainable = INTERACTION_SPECS[
+          oldState.interaction.id].is_string_classifier_trainable;
```
Why is this `is_string_classifier_trainable`? Should not we change it to more general, since it will be present for all trainable interactions.

pranavsid98 replied:
Done. Changed it to is_interaction_trainable.

prasanna08 commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -125,27 +127,19 @@ oppia.factory('AnswerClassificationService', [
+          var predictionService = (
+            PredictionAlgorithmRegistryService.getPredictionService(
+              algorithmId));
+          deferred.resolve(predictionService.predict(classifierData, answer));
```
This is incorrect. The `predict` method will only return the answer group index of the predicted outcome. You still have to prepare following object 
```
deferred.resolve({
    outcome: result.outcome,
    ruleIndex: result.rule_spec_index,
    answerGroupIndex: result.answer_group_index,
    classificationCategorization: result.classification_categorization
});
```
Once you have output from prediction service use it as answer group index and create above object. Predict won't have access to `rule_spec_index` and `outcome` associated with answer group index.

pranavsid98 replied:
Done. Regarding the deferred.resolve(), we are making it synchronous anyway right?

prasanna08 replied:
Nope. If you use normal returns then you will have to modify code where this function gets called.

prasanna08 commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -125,27 +127,19 @@ oppia.factory('AnswerClassificationService', [
+          var predictionService = (
+            PredictionAlgorithmRegistryService.getPredictionService(
+              algorithmId));
+          deferred.resolve(predictionService.predict(classifierData, answer));
```
This is incorrect. The `predict` method will only return the answer group index of the predicted outcome. You still have to prepare following object 
```
deferred.resolve({
    outcome: result.outcome,
    ruleIndex: result.rule_spec_index,
    answerGroupIndex: result.answer_group_index,
    classificationCategorization: result.classification_categorization
});
```
Once you have output from prediction service use it as answer group index and create above object. Predict won't have access to `rule_spec_index` and `outcome` associated with answer group index.

prasanna08 replied:
Also do not resolve directly. What if the prediction service's answer is `undefined`? That can happen in code classifier if IE11 is used as browser. So it'd be good to check the prediction before passing it.

prasanna08 commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationServiceSpec.js:
```diff
@@ -336,28 +349,18 @@ describe('Answer classification service with string classifier enabled',
-        rule_spec_index: 0,
-        classification_categorization: EXPLICIT_CLASSIFICATION
-      };
+    it('should query the prediction service if no answer group matches and ' +
```
What does this test do exactly? I am unable to understand. It says that 'it should query prediction service' but you are still comparing it with default outcome classification result. And nothing else indicates that prediction service has been called or something?

pranavsid98 replied:
@prasanna08 The predictionService is being called. My sample service just returns the default values for now. Check Line 360

prasanna08 replied:
I am still unable to understand this. Even if your sample service returns default outcome as of now the `classificationCategorization` should be different than what it is now. Also check my previous comment on what to return when we use ML classifier and what to return in case of default outcome.

pranavsid98 replied:
@prasanna08 Ahh. Thats fine. I wasn't very sure about what gets returned when we use the ML classifier.

pranavsid98 replied:
Done.

prasanna08 commented on feconf.py:
```diff
@@ -114,10 +114,10 @@
 
 # The minimum number of training samples required for training a classifier.
-MIN_TOTAL_TRAINING_EXAMPLES = 50
+MIN_TOTAL_TRAINING_EXAMPLES = 5
```
Why modify this?

pranavsid98 replied:
@prasanna08 So, to test handlers, we need to load demo explorations. Only one demo exploration has trainable states and it has like 10 training examples. The previous number of min_total_examples was just some random number we put up. When this actually gets into prod, we will decide on a concrete value for this.

prasanna08 replied:
I see. I don't know how such case should be handled. Normally on backend we swap variables but here it won't be possible. @seanlip do you have any idea for this?

seanlip replied:
If you're writing a backend test, then you can use self.swap as @prasanna08 says.

If this is just for convenience of testing locally, then change it when testing locally and change it back to the intended value in production before submitting to GitHub. The value here should be the correct one, or as near to it as possible. If we don't know the correct value then we should figure out the correct value.

pranavsid98 replied:
Done. Used self.swap().

prasanna08 commented on core/templates/dev/head/pages/exploration_player/PredictionAlgorithmSampleServiceSpec.js:
```diff
@@ -0,0 +1,43 @@
+ */
+
+ describe('Prediction algorithm sample service', function() {
+   beforeEach(module(oppia));
```
This is the problem with test. Oppia should be enclosed within quotes.

pranavsid98 replied:
Ahh. Thanks! Done.

prasanna08 reviewed (changes requested):
Took a pass! Overall structure looks good.

pranavsid98 committed:
changes

pranavsid98 committed:
fixed front end tests

pranavsid98 committed:
Merge branch 'develop' into front_end_framework

pranavsid98 committed:
fixed merge changes

pranavsid98 committed:
minor changes

pranavsid98 committed:
minor changes

pranavsid98 committed:
executing e2e tests

pranavsid98 committed:
fix lint errors

pranavsid98 committed:
fixed lint issues

pranavsid98 commented:
Whew! The flow is implemented and e2e tests work well! Thanks @prasanna08 
@seanlip @AllanYangZhou . I think this PR is ready for another round of reviewing. There might be some structural changes required in terms of where to keep these files and the like.

@anmolshkl It would be great if you take a final look too!

Thanks!!!

pranavsid98 commented on core/templates/dev/head/pages/exploration_editor/exploration_editor.html:
```diff
@@ -279,6 +279,10 @@
   <script src="{{TEMPLATE_DIR_PREFIX}}/pages/exploration_player/AnswerClassificationService.js"></script>
+  <script src="{{TEMPLATE_DIR_PREFIX}}/pages/exploration_player/StateClassifierMappingService.js"></script>
+  <script src="{{TEMPLATE_DIR_PREFIX}}/pages/exploration_player/PredictionAlgorithmRegistryService.js"></script>
+  <script src="{{TEMPLATE_DIR_PREFIX}}/pages/exploration_player/PredictionAlgorithmSampleService.js"></script>
```
This part is slightly tricky. @seanlip Is there anyway we can do dynamic inclusion here? Like only include the service that is returned by the registry service?

seanlip replied:
Not straightforwardly. I think it's OK to do this for now since it's all needed in the editor anyway. But in the learner view, please put a TODO to only load the algorithms that are needed.

pranavsid98 replied:
Done.

prasanna08 replied:
@seanlip I have one idea. We can incorporate these (prediction service) dependencies similar to how interaction dependencies are added. We will add prediction service dependencies in `interaction.py` file. Then we can add a helper function in `interaction_registery` which will get us deduplicated prediction service dependencies. We can store all these dependencies in `extensions/dependencies/prediction_service_dependencies`, from where we will get their respective html files and directly include them in frontend. (The html files will contain bunch of `<script src=""></script>` tags.)
How does this scheme sound? If its okay then I will start working on it in a separate PR.

seanlip replied:
Oh, I like that. That sounds good to me. Thanks!

prasanna08 replied:
OK, I will start working on it then but I think this PR should go in before I work on that, otherwise there will be complex merging issues.

seanlip commented on core/controllers/reader.py:
```diff
@@ -213,6 +213,24 @@ def get(self, exploration_id):
+                algorithm_id = classifier_training_jobs[index].algorithm_id
+                state_classifier_mapping[state_name] = {
+                    'algorithm_id': algorithm_id,
+                    'classifier_data': classifier_data
```
Is data schema version needed? /cc @anmolshkl @prasanna08

pranavsid98 replied:
@seanlip It's necessity comes down to this. Lets say the classifier for TextInput has a data_schema_version increment. This in turn changes the TextClassifierPredictionService.predict() method.
So, we can either maintain each data_schema_version's corresponding prediction service or just make the prediction service use the latest schema. 
I am slightly inclined towards the first option when I think about it. If we do end up doing that, we'll pass data_schema_version in this as well (and modify RegistryService accordingly).

seanlip replied:
Probably the first option sounds better, yep.

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -125,27 +127,30 @@ oppia.factory('AnswerClassificationService', [
-        if (result.outcome === defaultOutcome &&
-            INTERACTION_SPECS[oldState.interaction.id]
-              .is_string_classifier_trainable &&
+        var ruleBasedOutcomeIsDefault = (result.outcome == defaultOutcome);
```
===

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -125,27 +127,30 @@ oppia.factory('AnswerClassificationService', [
+            var predictedAnswerGroupIndex = predictionService.predict(
+              classifierData, answer).answerGroupIndex;
             deferred.resolve({
               outcome: result.outcome,
```
Could we make a domain object (AnswerClassificaitonResult, perhaps) for this "thing"? We're starting to duplicate its structure in multiple places.

pranavsid98 replied:
That would definitely make this cleaner. I'll start work on that.

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/PlayerServices.js:
```diff
@@ -297,79 +301,80 @@ oppia.factory('oppiaPlayerService', [
         answerIsBeingProcessed = true;
-        var oldState = exploration.getState(
-          playerTranscriptService.getLastStateName());
+        var stateName = playerTranscriptService.getLastStateName();
```
Perhaps oldStateName

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/PlayerServices.js:
```diff
@@ -297,79 +301,80 @@ oppia.factory('oppiaPlayerService', [
-          }
+          _explorationId, stateName, oldState, answer, false,
+          interactionRulesService)
+          .then(function(classificationResult) {
```
Bring the ')' from the previous line down to the start of this line (similar to how it was before), then you don't need all the indentation changes.

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/PredictionAlgorithmRegistryService.js:
```diff
@@ -0,0 +1,33 @@
+     // This mapping needs to be updated whenever a new prediction service needs
+     // to be added for classification.
+     var algorithmIdPredictionServiceMapping = {
+       LDAStringClassifier: PredictionAlgorithmSampleService
```
So, I guess we shouldn't ship this in production (or to develop, for that matter...)

I think for now you'll need to use {} here, and, only in the Karma tests, inject the sample service. Code going into develop must be clean and ready to be shipped to production at any time.

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/PredictionAlgorithmSampleService.js:
```diff
@@ -0,0 +1,27 @@
+
+ oppia.factory('PredictionAlgorithmSampleService', [function() {
+   return {
+     predict: function(classifierData, userText) {
```
Why call this userText and not answer (or learnerAnswer)?

pranavsid98 replied:
Done. changed it to answer.

seanlip commented on core/templates/dev/head/pages/exploration_player/StateClassifierMappingService.js:
```diff
@@ -0,0 +1,43 @@
+     },
+
+     getClassifierData: function(stateName) {
+       if (stateClassifierMapping[stateName]) {
```
if (stateClassifierMapping.hasOwnProperty(stateName))

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/StateClassifierMappingService.js:
```diff
@@ -0,0 +1,43 @@
+     },
+
+     getAlgorithmId: function(stateName) {
+       if (stateClassifierMapping[stateName]) {
```
Ditto

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationServiceSpec.js:
```diff
@@ -24,7 +24,7 @@ describe('Answer classification service with string classifier disabled',
       module(function($provide) {
         $provide.constant('INTERACTION_SPECS', {
           RuleTest: {
-            is_string_classifier_trainable: false
```
Careful. There are other instances of is_string_classifier_trainable in the backend codebase that also need to be replaced.

pranavsid98 replied:
Fixed.

seanlip reviewed (commented):
Hi @pranavsid98, thanks! I've taken a quick look through, and this looks quite reasonable. But really @prasanna08 and @AllanYangZhou should be reviewing -- especially @prasanna08, since he's going to be using the system you're setting up -- so I defer to them for further passes. I'm happy to take a look once more just before this is ready to go in.

pranavsid98 committed:
addresed one set of review comments

prasanna08 commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -125,27 +127,35 @@ oppia.factory('AnswerClassificationService', [
+                algorithmId));
+            var predictedAnswerGroupIndex;
+            if (predictionService) {
+              predictedAnswerGroupIndex = predictionService.predict(
```
This should be `predictedAnswerGroupIndex = predictionService.predict(classifier_data, answer);`. The predict function will return a singular value and not an object.

pranavsid98 replied:
Done.

prasanna08 commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -125,27 +127,35 @@ oppia.factory('AnswerClassificationService', [
+              predictedAnswerGroupIndex = result.answerGroupIndex;
+            }
             deferred.resolve({
               outcome: result.outcome,
```
So once you get predictedAnswerGroupIndex do this (you will need to do this separately in **if** part and **else** part),
```
outcome: answerGroups[predictedAnswerGroup].outcome,
answerGroupIndex: predictedAnswerGroup,
ruleIndex: <iterate through rules to find fuzzy rules index>
classificationCategorization: FUZZY_RULES_CLASSIFICATION (or something similar)
```

Finding `ruleIndex`
```
for (rule_spec in answerGroups[predictedAnswerGroupIndex].rule_specs) {
    if (rule_spec.rule_type === RULE_TYPE_CLASSIFIER) {
        ruleIndex = classifier_rule_spec_index;
        break;
    }
}
```

Does that help understanding what I'm trying to say?

pranavsid98 replied:
Yep. Thanks for explaining it so briefly. Done.

prasanna08 commented on core/templates/dev/head/pages/exploration_player/PredictionAlgorithmSampleService.js:
```diff
@@ -0,0 +1,27 @@
+ oppia.factory('PredictionAlgorithmSampleService', [function() {
+   return {
+     predict: function(classifierData, answer) {
+       return {
```
Only `return 0;`

pranavsid98 replied:
Done.

prasanna08 commented on core/templates/dev/head/pages/exploration_player/PredictionAlgorithmRegistryService.js:
```diff
@@ -0,0 +1,39 @@
+           return null;
+         }
+       },
+       setMapping: function(newAlgorithmIdPredictionServiceMapping) {
```
Do we need this somewhere?

pranavsid98 replied:
@prasanna08. Yes.The mapping variable inside the registry service has to be empty and not use testing sample services when we push to develop. So, we use this setMapping method while writing Karma tests.

prasanna08 reviewed (commented):
Hi @pranavsid98 took a pass. There are still some flaws in flow.

pranavsid98 committed:
addressed all review comments

pranavsid98 commented:
Hi @seanlip @prasanna08 . I've addressed all the changes:

-  made a Domain object for the AnswerClassificationResult object.

- Included dataSchemaVersion when passing from backend and use it to retrieve predictionService. This required a slight modification in the structure of the PredictionAlgorithmRegistryService mapping (apparently having objects/tuples as object properties isn't possible in JS?). I ended up mapping algorithmId to a list of objects where each object contains dataSchemaVersion and predictionService as properties. Is that fine?

- Also changed structure of how result gets returned when Classification happens.

Thanks!

pranavsid98 committed:
fix e2e tests

seanlip commented:
Prasanna - PTAL at this PR and give lgtm if you are happy (or add follow-up
comments if you are not)? I will probably only take a look after you are
happy with it.

Thanks!

On Monday, August 14, 2017, Prasanna Patil <notifications@github.com> wrote:

> *@prasanna08* commented on this pull request.
> ------------------------------
>
> In core/templates/dev/head/pages/exploration_editor/
> exploration_editor.html
> <https://github.com/oppia/oppia/pull/3716#discussion_r132943527>:
>
> > @@ -279,6 +279,10 @@
>    <script src="{{TEMPLATE_DIR_PREFIX}}/pages/exploration_player/ExplorationPlayerStateService.js"></script>
>    <script src="{{TEMPLATE_DIR_PREFIX}}/pages/exploration_player/PlayerConstants.js"></script>
>    <script src="{{TEMPLATE_DIR_PREFIX}}/pages/exploration_player/AnswerClassificationService.js"></script>
> +  <script src="{{TEMPLATE_DIR_PREFIX}}/pages/exploration_player/StateClassifierMappingService.js"></script>
> +  <script src="{{TEMPLATE_DIR_PREFIX}}/pages/exploration_player/PredictionAlgorithmRegistryService.js"></script>
> +  <script src="{{TEMPLATE_DIR_PREFIX}}/pages/exploration_player/PredictionAlgorithmSampleService.js"></script>
>
> OK, I will start working on it then but I think this PR should go in
> before I work on that, otherwise there will be complex merging issues.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/3716#discussion_r132943527>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKFeyrL36c2i9y8WkNupxIp5f-cYxQhEks5sYERTgaJpZM4OvDWZ>
> .
>

pranavsid98 commented:
Wow. This is weird. All e2e tests and front-end tests execute on my local dev machine.

Edit: Ouch. Forgot to do 'git add' when I added domain file for AnswerClassificationResult. Resolved.

pranavsid98 committed:
fix tests

pranavsid98 committed:
removed log file - added by mistake

prasanna08 commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -61,30 +58,32 @@ oppia.factory('AnswerClassificationService', [
       }
     };
 
+    var findRuleIndex = function(answerGroup, answerGroupIndex) {
```
findClassifierRuleIndex or something similar?

pranavsid98 replied:
Done.

prasanna08 commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -125,27 +118,42 @@ oppia.factory('AnswerClassificationService', [
+                predictedAnswerGroupIndex,
+                findRuleIndex(answerGroups[predictedAnswerGroupIndex],
+                              predictedAnswerGroupIndex),
+                TRAINING_DATA_CLASSIFICATION
```
MACHINE_LEARNING_CLASSIFICATION or something. We aren't using training data but trained model for classification. So something along trained models and machine learning classifiers.

pranavsid98 replied:
Done. Created a new constant for that. What is statistical classification btw? Is it used anywhere?

seanlip replied:
Actually I think "statistical classification" is our current term for ML classification, so I suggest using that if possible.

prasanna08 reviewed (approved):
Looks good. Just a couple of neet comments.

seanlip commented on core/controllers/reader_test.py:
```diff
@@ -188,6 +193,51 @@ def test_give_feedback_handler(self):
+
+        # Load demo exploration.
+        exp_services.delete_demo(exploration_id)
+        with self.swap(feconf, 'ENABLE_ML_CLASSIFIERS', True):
```
In what way does all this stuff actually affect the call to load_demo()?

pranavsid98 replied:
@seanlip load_demo() calls save_exploration_from_yaml() method. Only if ML classifiers are enabled, that function will lead to training job creation for trainable states.

seanlip replied:
Ah, OK. Consider adding that info as a comment inline.

Do you need the other self.swaps? If so, justify them similarly; if not, please remove them. Thanks!

pranavsid98 replied:
We do. I've added comments for justifying them.

seanlip commented on core/controllers/reader_test.py:
```diff
@@ -188,6 +193,51 @@ def test_give_feedback_handler(self):
         self.logout()
 
 
+class ExplorationHandlerTest(test_utils.GenericTestBase):
```
This is a bit vague/non-specific. Perhaps ExplorationStateClassifierMappingTests?

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -61,30 +58,32 @@ oppia.factory('AnswerClassificationService', [
+          return i;
+        }
+      }
+    };
```
What do you want to return if it's not found? Null? -1?

Consider adding a short docstring above the function to explain what the function does and what convention you use here.

pranavsid98 replied:
@seanlip I've added a doctoring above the function. About not being found, I don't really think thats possible until there is a bug (maybe if the predicted answer group index might be incorrect) - so we should return null for that. If the prediction is correct, the predicted answer group must have a rule_type RULE_TYPE_CLASSIFIER.

seanlip commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -125,27 +118,42 @@ oppia.factory('AnswerClassificationService', [
+            var predictionService = (
+              PredictionAlgorithmRegistryService.getPredictionService(
+                algorithmId, dataSchemaVersion));
+            var predictedAnswerGroupIndex;
```
Why not define this within the if block, if you don't use it outside? In fact I don't even think you need this line, you can just put "var" before predictedAnswerGroupIndex a few lines below.

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -61,30 +58,32 @@ oppia.factory('AnswerClassificationService', [
       }
     };
 
+    var findRuleIndex = function(answerGroup, answerGroupIndex) {
```
You don't use answerGroupIndex. Why pass it in?

pranavsid98 replied:
Oh. I had modified it from passing in the answerGroups to just the specific answerGroup. Index is removed now. Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -125,27 +118,42 @@ oppia.factory('AnswerClassificationService', [
+                TRAINING_DATA_CLASSIFICATION
+              );
+            } else {
+              var classificationResult = result;
```
Seems weird to give it this name, especially since classification didn't happen...

Why not assign the thing on line 143 to the `result` variable, get rid of this else block, get rid of the other else block below and just do `deferred.resolve(result);` at the end?

pranavsid98 replied:
Done. I had initially kept it as the same variable, but Allan thought it might be a good idea to differentiate the result of classification from the other result.

seanlip commented on core/templates/dev/head/pages/exploration_player/PredictionAlgorithmRegistryService.js:
```diff
@@ -0,0 +1,60 @@
+    * to be added for classification. The mapping is from algorithmId to a
+    * list of objects. The mapping should be of the type:
+    * {
+    * algorithmId: [{
```
Indent this and the next three lines by 2 (after the asterisk).

seanlip replied:
Also indent this entire comment block by 1 so that the lowermost asterisk lines up with the var just below it.

pranavsid98 replied:
Done,

seanlip commented on core/templates/dev/head/pages/exploration_player/PredictionAlgorithmRegistryService.js:
```diff
@@ -0,0 +1,60 @@
+     return {
+       getPredictionService: function(algorithmId, dataSchemaVersion) {
+         if (algorithmIdPredictionServiceMapping.hasOwnProperty(algorithmId)) {
+           for (var i = 0; i < algorithmIdPredictionServiceMapping[
```
Wouldn't it be cleaner to just map the dataSchemaVersion to the predictionService (for each algorithmId)?

pranavsid98 replied:
Actually, yes! We can't use integers as properties, but dataSchemaVersion.toString() should work just fine. Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/PredictionAlgorithmRegistryService.js:
```diff
@@ -0,0 +1,60 @@
+       },
+       // The below function is required for running tests with sample
+       // prediction services.
+       setMapping: function(newAlgorithmIdPredictionServiceMapping) {
```
Just wondering -- outside tests, how do you plan to set the mapping? Will it be hardcoded here, or will it need to be a constant that the backend can also see?

pranavsid98 replied:
@seanlip Well, since this is the registry service, we could hardcore it here. Do you think it might be a better idea to keep it as a constant and use that constant here?

seanlip replied:
Don't know. If it's being used elsewhere then a (reusable) constant sounds good. If not then hardcoding here is fine.

pranavsid98 replied:
AFAIK its not being used anywhere else. The registry service is the only service that handles this mapping. Lets keep it hardcoded here for now then.

seanlip commented on core/templates/dev/head/pages/exploration_player/exploration_player.html:
```diff
@@ -163,6 +164,12 @@ <h1 class="oppia-exploration-h1"><span class="protractor-test-exploration-header
+  <script src="{{TEMPLATE_DIR_PREFIX}}/pages/exploration_player/PredictionAlgorithmRegistryService.js"></script>
+
+  <!-- TODO(pranavsid98): Load only required prediction services. -->
+  <script src="{{TEMPLATE_DIR_PREFIX}}/pages/exploration_player/PredictionAlgorithmSampleService.js"></script>
```
This shouldn't be loaded here.

pranavsid98 replied:
Depends on answer to comment below.

seanlip commented on core/templates/dev/head/pages/exploration_player/PredictionAlgorithmSampleService.js:
```diff
@@ -0,0 +1,25 @@
+ * @fileoverview Prediction Algorithm sample for testing framework.
+ */
+
+ oppia.factory('PredictionAlgorithmSampleService', [function() {
```
I don't think this should exist outside a non-Spec file. Can you incorporate it within the karma test file somehow?

pranavsid98 replied:
@seanlip Doing that should be possible, but this is currently being used in a lot of tests (AnswerClassificationSpec and PredictionAlgorithmRegistrySpec). The whole Sample Service stuff can be done away with the second we implement our first actual prediction service. Seeing that Prasanna's PR is already almost done, I think it would be better to deprecate all of this in his PR. If pushing testing services to develop is an issue, would it be a better idea to merge this PR with Prasanna;s PR when both PRs are good and ready to be merged?

seanlip replied:
Nope. We should have a "simple" sample service for testing in any case, regardless of whatever is being used in production, because the production services are going to be quite complex and not suitable for debugging pipeline issues that arise in testing. You can probably provide a custom service in tests using $provide.factory().

pranavsid98 replied:
Sounds good. Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/PredictionAlgorithmSampleServiceSpec.js:
```diff
@@ -0,0 +1,38 @@
+ * @fileoverview Unit tests for the sample prediction algorithm service.
+ */
+
+ describe('Prediction algorithm sample service', function() {
```
You don't need this test; the "sample service" is not an actual thing.

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/StateClassifierMappingService.js:
```diff
@@ -0,0 +1,51 @@
+
+   return {
+     init: function(newStateClassifierMapping) {
+       stateClassifierMapping = newStateClassifierMapping;
```
Better to transfer to JS names and a domain object:

```
    stateClassifierMapping = {};
    for (var stateName in backendStateClassifierMapping) {
      stateClassifierMapping[stateName] = ClassifierObjectFactory.create(
        algorithmId, classifierData, ...);
    }
```

pranavsid98 replied:
Done. Good point on making this a ClassifierObjectFactory class.

seanlip commented on core/templates/dev/head/pages/exploration_player/StateClassifierMappingService.js:
```diff
@@ -0,0 +1,51 @@
+ * @fileoverview Services for mapping state names to classifier details.
+ */
+
+ oppia.factory('StateClassifierMappingService', [function() {
```
Tests?

pranavsid98 replied:
Done.

seanlip reviewed (changes requested):
Looks pretty good! Needs a little bit more work though, I think.

seanlip commented:
Btw, just a quick question: will answer classification become synchronous in the future? Is there an issue to track this?

(Asking because this affects some stuff @kevintab95 is doing.)

pranavsid98 committed:
addressed review changes

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -99,6 +99,9 @@
 STATISTICAL_CLASSIFICATION = 'statistical_classifier'
+# Represents answers which were predicted using a machine learning model from
+# training data within an answer group.
+MACHINE_LEARNING_CLASSIFICATION = 'ml_classifier'
```
Don't define a new thing here; this is the same as statistical_classification so please use that instead.

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -24,18 +24,19 @@ oppia.constant('EXPLICIT_CLASSIFICATION', 'explicit')
 oppia.constant('TRAINING_DATA_CLASSIFICATION', 'training_data_match')
 oppia.constant('STATISTICAL_CLASSIFICATION', 'statistical_classifier')
 oppia.constant('DEFAULT_OUTCOME_CLASSIFICATION', 'default_outcome')
+oppia.constant('MACHINE_LEARNING_CLASSIFICATION', 'ml_classifier')
```
Ditto -- drop the new constant, here and below.

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -75,13 +76,23 @@ oppia.factory('AnswerClassificationService', [
+    var findClassifierRuleIndex = function(answerGroup) {
       for (var i = 0; i < answerGroup.rules.length; i++) {
         var rule = answerGroup.rules[i];
         if (rule.type == RULE_TYPE_CLASSIFIER) {
```
===

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -75,13 +76,23 @@ oppia.factory('AnswerClassificationService', [
           return i;
         }
       }
+      return null;
```
throw Error('...');

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationServiceSpec.js:
```diff
@@ -241,7 +241,7 @@ describe('Answer classification service with string classifier enabled',
     var EXPLICIT_CLASSIFICATION = 'explicit';
     var DEFAULT_OUTCOME_CLASSIFICATION = 'default_outcome';
-    var TRAINING_DATA_CLASSIFICATION = 'training_data_match';
+    var MACHINE_LEARNING_CLASSIFICATION = 'ml_classifier';
```
ditto

seanlip replied:
though shouldn't you be importing the existing constants rather than redefining new ones here? (I realize this was a fault of the previous tests, but might be worth fixing it.)

pranavsid98 replied:
True. Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/PredictionAlgorithmRegistryService.js:
```diff
@@ -20,32 +20,22 @@
-             }
-           }
+           return algorithmIdPredictionServiceMapping[algorithmId][
+             dataSchemaVersion.toString()]
```
missing semicolon at end.

pranavsid98 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/PredictionAlgorithmRegistryService.js:
```diff
@@ -20,32 +20,22 @@
-               return null;
-             }
-           }
+           return algorithmIdPredictionServiceMapping[algorithmId][
```
Please add a comment above this line to explain why we're converting dataSchemaVersion to a string. (I think it's a good idea but this is just for clarity for future devs.)

pranavsid98 replied:
Done.

pranavsid98 committed:
one round of review comments addressed

pranavsid98 committed:
modified StateClassifierMappingService

pranavsid98 committed:
fixed e2e tests

seanlip commented on core/templates/dev/head/pages/exploration_player/StateClassifierMappingService.js:
```diff
@@ -0,0 +1,50 @@
+         stateClassifierMapping = {};
+         var algorithmId, classifierData, dataSchemaVersion;
+         for (var stateName in backendStateClassifierMapping) {
+           if (backendStateClassifierMapping[stateName]) {
```
Should be: backendStateClassifierMapping.hasOwnProperty(stateName)

pranavsid98 replied:
Done. Should have done this myself. Still getting used to the JS way of checking conditions :)

seanlip reviewed (approved):
This looks great! Just one minor comment, and then it's fine to merge.

@prasanna08 -- if you are blocking on this PR, feel free to merge this and fix the comment later.

Thanks!

pranavsid98 committed:
final change addressed

pranavsid98 commented:
Thanks @seanlip . Addressed the final comment.

seanlip commented:
Hm, lots of test failures. Any idea why?

pranavsid98 commented:
@seanlip Ahh. Every stateName is present in this mapping. It's mapped to null if classifier does not exist. So hasOwnProperty(stateName) is True always. Might be a better idea to not have the state names which don't have classifiers in the mapping.

pranavsid98 committed:
fix e2e tests

seanlip commented:
Yep, agreed!
